src/segment_strava.go

```diff
@@ -0,0 +1,222 @@
+	StravaRefreshToken = "strava_refresh_token"
+)
+
+const (
+	RideIcon            Property = "ride_icon"
+	RunIcon             Property = "run_icon"
+	SkiingIcon          Property = "skiing_icon"
+	WorkOutIcon         Property = "workout_icon"
+	UnknownActivityIcon Property = "unknown_activity_icon"
+)
+
+// StravaData struct contains the API data
+type StravaData struct {
+	Type      string    `json:"type"`
+	StartDate time.Time `json:"start_date"`
+	Name      string    `json:"name"`
+	Distance  float64   `json:"distance"`
+	Duration  float64   `json:"moving_time"`
+}
+
+type TokenExchange struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    int    `json:"expires_in"`
+}
+
+type AuthError struct{}
+
+func (a *AuthError) Error() string {
+	return "unable to authenticate"
+}
+
+func (s *strava) enabled() bool {
+	data, err := s.getResult()
+	if err == nil {
+		s.StravaData = *data
+		s.ActivityIcon = s.getActivityIcon()
+		s.Hours = s.getHours()
+		s.Ago = s.getAgo()
+		return true
+	}
+	if _, s.Authenticate = err.(*AuthError); s.Authenticate {
+		// TODO: should prompt the user to (re)authenticate
+		return true
+	}
+	return false
+}
+
+func (s *strava) getHours() int {
+	hours := time.Since(s.StartDate).Hours()
+	return int(math.Floor(hours))
+}
+
+func (s *strava) getAgo() string {
+	if s.Hours > 24 {
+		days := int32(math.Floor(float64(s.Hours) / float64(24)))
+		return fmt.Sprintf("%d", days) + string('d')
+	}
+	return fmt.Sprintf("%d", s.Hours) + string("h")
+}
+
+func (s *strava) getActivityIcon() string {
+	switch s.Type {
+	case "Ride":
+		return s.props.getString(RideIcon, "\uf5a2")
+	case "Run":
+		return s.props.getString(RunIcon, "\ufc0c")
+	case "NordicSki":
+	case "AlpineSki":
+	case "BackcountrySki":
+		return s.props.getString(SkiingIcon, "\ue213")
+	case "WorkOut":
+		return s.props.getString(WorkOutIcon, "\ue213")
+	default:
+		return s.props.getString(UnknownActivityIcon, "\ue213")
+	}
+	return s.props.getString(UnknownActivityIcon, "\ue213")
+}
+
+func (s *strava) string() string {
+	segmentTemplate := s.props.getString(SegmentTemplate, "{{ .Ago }}")
+	template := &textTemplate{
+		Template: segmentTemplate,
+		Context:  s,
+		Env:      s.env,
+	}
+	text, err := template.render()
+	if err != nil {
+		return err.Error()
+	}
+	return text
+}
+
+func (s *strava) getAccessToken() (string, error) {
+	// get directly from cache
+	if acccessToken, OK := s.env.cache().get(StravaAccessToken); OK {
+		return acccessToken, nil
+	}
+	if refreshToken, OK := s.env.cache().get(StravaRefreshToken); OK {
+		if acccessToken, err := s.refreshToken(refreshToken); err == nil {
+			return acccessToken, nil
+		}
+		return "", &AuthError{}
+	}
+	// use initial tokens from property
+	accessToken := s.props.getString(AccessToken, "")
+	refreshToken := s.props.getString(RefreshToken, "")
+	if len(accessToken) == 0 || len(refreshToken) == 0 {
+		return "", &AuthError{}
+	}
```

Comment:
@JanDeDobbeleer I think we should take one extra attempt to call refreshToken() to get an accesstoken from the initial refresh token before we givue up due to the following scenario:

* Refreshtoken stored in cache is invalid for some reason
* User has entered a new valid refreshtoken in the props

Reply:
Fair point!